builtin/providers/aws/resource_aws_alb_listener.go

```diff
@@ -0,0 +1,255 @@
+
+			"protocol": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  "HTTP",
+				StateFunc: func(v interface{}) string {
+					return strings.ToUpper(v.(string))
+				},
+				ValidateFunc: validateAwsAlbListenerProtocol,
+			},
+
+			"ssl_policy": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
+			"certificate_arn": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
+			"default_action": {
+				Type:     schema.TypeList,
+				Required: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"target_group_arn": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"type": {
+							Type:         schema.TypeString,
+							Required:     true,
+							ValidateFunc: validateAwsAlbListenerDefaultActionType,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func resourceAwsAlbListenerCreate(d *schema.ResourceData, meta interface{}) error {
+	elbconn := meta.(*AWSClient).elbv2conn
+
+	params := &elbv2.CreateListenerInput{
+		LoadBalancerArn: aws.String(d.Get("load_balancer_arn").(string)),
+		Port:            aws.Int64(int64(d.Get("port").(int))),
+		Protocol:        aws.String(d.Get("protocol").(string)),
+	}
+
+	if sslPolicy, ok := d.GetOk("ssl_policy"); ok {
+		params.SslPolicy = aws.String(sslPolicy.(string))
+	}
+
+	if certificateArn, ok := d.GetOk("certificate_arn"); ok {
+		params.Certificates = make([]*elbv2.Certificate, 1)
+		params.Certificates[0] = &elbv2.Certificate{
+			CertificateArn: aws.String(certificateArn.(string)),
+		}
+	}
+
+	if defaultActions := d.Get("default_action").([]interface{}); len(defaultActions) == 1 {
+		params.DefaultActions = make([]*elbv2.Action, len(defaultActions))
+
+		for i, defaultAction := range defaultActions {
+			defaultActionMap := defaultAction.(map[string]interface{})
+
+			params.DefaultActions[i] = &elbv2.Action{
+				TargetGroupArn: aws.String(defaultActionMap["target_group_arn"].(string)),
+				Type:           aws.String(defaultActionMap["type"].(string)),
+			}
+		}
+	}
+
+	resp, err := elbconn.CreateListener(params)
+	if err != nil {
+		return errwrap.Wrapf("Error creating ALB Listener: {{err}}", err)
+	}
+
+	if len(resp.Listeners) == 0 {
+		return errors.New("Error creating ALB Listener: no listeners returned in response")
+	}
+
+	d.SetId(*resp.Listeners[0].ListenerArn)
+
+	return resourceAwsAlbListenerRead(d, meta)
+}
+
+func resourceAwsAlbListenerRead(d *schema.ResourceData, meta interface{}) error {
+	elbconn := meta.(*AWSClient).elbv2conn
+
+	resp, err := elbconn.DescribeListeners(&elbv2.DescribeListenersInput{
+		ListenerArns: []*string{aws.String(d.Id())},
+	})
+	if err != nil {
+		if isListenerNotFound(err) {
+			log.Printf("[WARN] DescribeListeners - removing %s from state", d.Id())
+			d.SetId("")
+			return nil
+		}
+		return errwrap.Wrapf("Error retrieving Listener: {{err}}", err)
+	}
+
+	if len(resp.Listeners) != 1 {
+		return fmt.Errorf("Error retrieving Listener %q", d.Id())
```

Comment:
This is applying the pattern discussed with @phinze - if we are adding context to an error returned by someone else we wrap, but if constructing our own error we use fmt.Errorf (or errors.New if there are no format placeholders). This is a standard worth rolling out across the entire code base, but will take a very long time, hence introducing it with new resources and modifying existing ones over time.